extensions/amp-lightbox-gallery/1.0/component.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {Lightbox} from './../../amp-lightbox/1.0/lightbox';
+import {LightboxGalleryContext} from './context';
+import {sequentialIdGenerator} from '../../../src/utils/id-generator';
+import {
+  useContext,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from '../../../src/preact';
+
+const generateLightboxItemKey = sequentialIdGenerator();
+
+/**
+ * @param {!LightboxGalleryDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function LightboxGallery({children}) {
+  const lightboxRef = useRef(null);
+  const lightboxElements = useRef([]);
+  const register = (key, render) => {
+    lightboxElements.current[key] = render();
+  };
+  const deregister = (key) => {
+    delete lightboxElements.current[key];
+  };
+  const context = {
+    deregister,
+    register,
+    open: () => lightboxRef.current.open(),
+  };
+  return (
+    <>
+      <Lightbox ref={lightboxRef} scrollable>
+        {/* TODO: This needs an actual close button UI */}
+        <div
+          role="button"
+          tabindex="0"
+          onClick={() => lightboxRef.current.close()}
+        >
+          Close lightbox
+        </div>
+        <div>{lightboxElements.current}</div>
```

Comment:
Same question. IMHO it'd be better to wait until lightbox is opened. Perhaps lightbox itself can have a `render()` prop that's execute like this: `{render ? render() : children}`?

Reply:
These are not shown until the lightbox is opened because of the `Lightbox` internals: https://github.com/ampproject/amphtml/blob/46b225176dd40c05591132e5f61bd4f6ac44769f/extensions/amp-lightbox/1.0/lightbox.js#L153

However I did add a `renderers` ref who only called as needed given to the `Lightbox` `onBeforeOpen`

Reply:
Yeah. That's the distinction that: lightbox saves the DOM costs, but doing it higher up also saves JSX construction costs.

Reply:
Oh I see what you mean. Done.